HDDS-6153. Add simple integration test to the read-replicas debug tool.

dombizita opened the pull request:
## What changes were proposed in this pull request?

In this change I added acceptance robot tests to the previously added read-replicas debug CLI tool. There was an already existing ozone-debug.robot test which I made into a resource file with keywords to check the JSON file generated with read-replicas. I made an ozone-debug-tests.robot for the chunkinfo and read-replicas CLI tool related tests in basic environment. I added two test files, one for the case when one of the block replicas became corrupt and one when one of the datanodes is unavailable. I changed the block size in the compose/ozone docker environment, so I can test the read-replicas command with a key with more than one block and it won't take that much to execute it. I am executing the added tests in the compose/ozone’s test.sh via a method in testlib.sh.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-6153

## How was this patch tested?

Executed the related robot tests successfully.

sadanand48 commented on hadoop-ozone/dist/src/main/smoketest/debug/ozone-debug.robot:
```diff
@@ -15,22 +15,48 @@
-                    Should contain      ${result}       files
-    ${result} =     Execute             ozone debug chunkinfo o3://om/vol1/bucket1/debugKey | jq -r '.KeyLocations[0][0].Locations.files[0]'
-                    File Should Exist   ${result}
+Check cheksum mismatch error
```
typo : checksum

sadanand48 commented on hadoop-ozone/dist/src/main/smoketest/debug/ozone-debug-tests.robot:
```diff
@@ -0,0 +1,60 @@
+                    File Should Exist   ${result}
+
+Test ozone debug read-replicas
+    Execute                             ozone debug read-replicas o3://om/${VOLUME}/${BUCKET}/${TESTFILE}
```
First few checks/lines  are common in all three tests, Can we move to a common function?

dombizita replied:
Thanks, this is a good idea. I added a keyword in the ozone-debug to execute the read-replicas there and do the common checks.

sadanand48 commented on hadoop-ozone/dist/src/main/smoketest/debug/ozone-debug-modified-env.robot:
```diff
@@ -0,0 +1,47 @@
+Resource            ozone-debug.robot
+Test Timeout        5 minute
+*** Variables ***
+${VOLUME}           vol1
```
CI is probably failing because there is already another volume vol1 created by some other test . We could try changing volume name here .

sadanand48 reviewed (commented):
Thanks @dombizita for working on this nice tool. Just few minor comments inline.

dombizita commented:
> Thanks @dombizita for working on this nice tool. Just few minor comments inline.

Hi @sadanand48, thanks for taking your time to look at my patch. I addressed your comments and updated my PR.

sadanand48 commented on hadoop-ozone/dist/src/main/compose/testlib.sh:
```diff
@@ -407,3 +407,33 @@ prepare_for_runner_image() {
+
+  execute_robot_test datanode debug/ozone-debug-tests.robot
+
+  docker exec ozone_datanode_2 find /data/hdds/hdds/ -name *.block | sed -n '2p' | xargs docker exec ozone_datanode_2 sed -i -e '1s/^/a/'
```
I think CI failed this time because maybe there are other keys created from other tests, so this might not be the right block file you are trying to corrupt.

sadanand48 replied:
Alternatively you could try this tool to get the exact file .
`ozone debug chunkinfo ${VOLUME}/${BUCKET}/${TESTFILE} |  jq -r '.KeyLocations[0][0].Locations.files[0]`

dombizita replied:
Yes, that was the problem, it wasn't the best way to do the corruption. Thanks, I fixed it with the use of chunkinfo.

sadanand48 commented on hadoop-ozone/dist/src/main/smoketest/debug/ozone-debug.robot:
```diff
@@ -15,22 +15,52 @@
-    Execute             ozone sh key put o3://om/vol1/bucket1/debugKey /opt/hadoop/NOTICE.txt
+Execute read-replicas CLI tool
+    Execute                         ozone debug read-replicas o3://om/${VOLUME}/${BUCKET}/${TESTFILE}
+    ${directory} =                  Execute     find /opt/hadoop -maxdepth 1 -name '${VOLUME}_${BUCKET}_${TESTFILE}_*' | tail -n 1
```
I think the output here should be sorted  based on timestamp so that the right directory is obtained.

dombizita replied:
Even though the directories are sorted by the creation time if they are ordered by name (the directories end with a timestamp), I looked after and find is not always consistent with the order. I changed it to use ls as it will be consistent with the order: "Sort entries alpha‐betically if none of -cftuvSUX nor --sort is specified.".
But I think the CI didn't fail because of this. The stopped datanode didn't become STALE and DEAD as the interval were 5 and 10 minutes. I changed that and added a test case for the STALE and one for the DEAD datanode case.

sadanand48 replied:
Thanks for clarifying @dombizita .

sadanand48 commented on hadoop-ozone/dist/src/main/smoketest/debug/ozone-debug-dead-datanode.robot:
```diff
@@ -0,0 +1,44 @@
+    ${datanodes_b1} =                   Create List   ${json}[blocks][0][replicas][0][hostname]    ${json}[blocks][0][replicas][1][hostname]
+    Check for datanodes                 ${datanodes_b1}    ${datanodes_expected}
+    ${datanodes_b2} =                   Create List   ${json}[blocks][1][replicas][0][hostname]    ${json}[blocks][1][replicas][1][hostname]
+    Check for datanodes                 ${datanodes_b2}    ${datanodes_expected}
```
'Check unavailable datanode error' could be added here as well .

dombizita replied:
I didn't add it, as it is checking if the manifest file contains the UNAVAILABLE error on that datanode, but that is only happening if the datanode is STALE. If the datanode is dead, there won't be an attempt to download from there, so it won't be appearing in the json. That is why the checking for datanodes only checking for two datanodes and the directory will contain only 5 files (one manifest file, 2x2 downloaded blocks).

sadanand48 replied:
Thanks for explaining.

sadanand48 commented:
Thanks @dombizita for the patch and addressing comments.  +1 LGTM .  Current CI failure looks unrelated to patch, please retrigger the CI.

sadanand48 commented on hadoop-ozone/dist/src/main/compose/ozone/test.sh:
```diff
@@ -50,6 +50,8 @@ execute_robot_test scm freon
 execute_robot_test scm cli
 execute_robot_test scm admincli
 
+execute_debug_tests
```
@dombizita Looks like the CI fails because of after restart of datanode 2 , scm is still in safe mode and the remaining tests do not wait for safemode exit, We can either move this test to the end or add a wait_for_safemode call here.

dombizita replied:
Hi @sadanand48, after some investigating I found what caused the failing. The admincli/pipeline.robot is creating a STANDALONE/ONE pipeline with one of the datanodes. After that in the debug tests I stop and reastart a datanode and I found that after that there won't be a RATIS/THREE pipeline, so the ozonfs tests are failing with the key uploading. It is happening because the pipeline limit is set to 1 on the datanodes and when we are trying to create a RATIS/THREE pipeline we are checking if the pipeline limit is reached on each datanode. When we are checking the current pipeline count on the datanodes we are deducting the single node pipelines, but we are only deducting the RATIS replication type ones. So we will count in the STANDALONE/ONE and with that we will reach the pipeline limit on that datanode, therefore we won't create the RATIS/THREE pipeline. I filed a JIRA (HDDS-6379) with a more detailed description.

fapifta reviewed (approved):
Thank you for working on this @dombizita, the changes look good to me, this is a great addition. I am merging it shortly.
